modules/kargoBidAdapter.js

```diff
@@ -4,286 +4,484 @@ import { registerBidder } from '../src/adapters/bidderFactory.js';
+    return hex.slice(0, 8) + '-' + hex.slice(8, 12) + '-' + hex.slice(12, 16) + '-' + hex.slice(16, 20) + '-' + hex.slice(20);
+  } catch (e) {
+    return '';
+  }
+}
+
+function _getCrb() {
+  let localStorageCrb = getCrbFromLocalStorage();
+  if (Object.keys(localStorageCrb).length) {
+    return localStorageCrb;
+  }
+  return getCrbFromCookie();
+}
+
+function _getSessionId() {
+  if (!sessionId) {
+    sessionId = _generateRandomUUID();
+  }
+  return sessionId;
+}
+
+function getCrbFromCookie() {
+  try {
+    const crb = JSON.parse(STORAGE.getCookie(CERBERUS.KEY));
+    if (crb && crb.v) {
+      let vParsed = JSON.parse(atob(crb.v));
+      if (vParsed) {
+        return vParsed;
       }
-      return {};
-    } catch (e) {
-      return {};
     }
-  },
+    return {};
+  } catch (e) {
+    return {};
+  }
+}
 
-  _getCrbFromLocalStorage() {
-    try {
-      return JSON.parse(atob(spec._getLocalStorageSafely('krg_crb')));
-    } catch (e) {
-      return {};
-    }
-  },
+function getCrbFromLocalStorage() {
+  try {
+    return JSON.parse(atob(getLocalStorageSafely(CERBERUS.KEY)));
+  } catch (e) {
+    return {};
+  }
+}
 
-  _getCrb() {
-    let localStorageCrb = spec._getCrbFromLocalStorage();
-    if (Object.keys(localStorageCrb).length) {
-      return localStorageCrb;
-    }
-    return spec._getCrbFromCookie();
-  },
-
-  _getLocalStorageSafely(key) {
-    try {
-      return storage.getDataFromLocalStorage(key);
-    } catch (e) {
-      return null;
-    }
-  },
-
-  _getUserIds(tdid, usp, gdpr) {
-    const crb = spec._getCrb();
-    const userIds = {
-      kargoID: crb.lexId,
-      clientID: crb.clientId,
-      crbIDs: crb.syncIds || {},
-      optOut: crb.optOut,
-      usp: usp
-    };
 
-    try {
-      if (gdpr) {
-        userIds['gdpr'] = {
-          consent: gdpr.consentString || '',
-          applies: !!gdpr.gdprApplies,
-        }
+
+function getLocalStorageSafely(key) {
+  try {
+    return STORAGE.getDataFromLocalStorage(key);
+  } catch (e) {
+    return null;
+  }
+}
+
+function getUserIds(tdid, usp, gdpr, eids) {
+  const crb = spec._getCrb();
+  const userIds = {
+    crbIDs: crb.syncIds || {}
+  };
+
+  if (tdid) {
+    userIds.tdID = tdid;
+  }
+
+  if(usp) {
+    userIds.usp = usp
+  }
+
+  try {
+    if (gdpr) {
+      userIds['gdpr'] = {
+        consent: gdpr.consentString || '',
+        applies: !!gdpr.gdprApplies,
       }
-    } catch (e) {
     }
-    if (tdid) {
-      userIds.tdID = tdid;
+  } catch (e) {
+  }
+  
+  if(crb.lexId != null) {
+    userIds.kargoID = crb.lexId
+  }
+
+  if(crb.clientId !=  null) {
+    userIds.clientID = crb.clientId
+  }
+
+  if(crb.optOut != null) {
+    userIds.optOut = crb.optOut
+  }
+
+  if(eids != null) {
+    userIds.eids = eids
+  }
+
+  return userIds;
+}
+
+function getClientId() {
+  const crb = spec._getCrb();
+  return crb.clientId;
+}
+
+function getAllMetadata(bidderRequest) {
+  return {
+    pageURL: bidderRequest?.refererInfo?.page,
```

Comment:
just curious what the question marks stand for?

Reply:
I wasnt sure the answer to this as this code was already present. Turns out this is `optional chaining` which prevents the throwing of errors if there are null values. Instead undefined is returned.
https://www.javascripttutorial.net/es-next/javascript-optional-chaining-operator/

Reply:
thanks for clarifying this one @njflynn